### PR TITLE
Scrum 81 connection requests relationships

### PR DESCRIPTION
**Pull request by K444RU**
1. Connection Request Endpoints

Send Connection Request:
Endpoint: POST /connections/requests/{userId}
Logic: ConnectionService.sendConnectionRequest handles the request, checking for self-requests, existing connections, and duplicates, then creates a new pending connection if valid.

Accept Connection Request:
Endpoint: PATCH /connections/requests/{requestId}/accept
Logic: ConnectionService.acceptConnectionRequest ensures the request is pending and the user is authorized, then updates the state to ACCEPTED.

Reject Connection Request:
Endpoint: PATCH /connections/requests/{requestId}/reject
Logic: ConnectionService.rejectConnectionRequest verifies the pending state and authorization, then marks it as REJECTED.

2. Get Connections Endpoint
Endpoint: GET /connections
Logic: ConnectionService.getConnections fetches all connections for a user and organizes them into:
Active: Accepted connections.
Pending Incoming: Requests sent to the user.
Pending Outgoing: Requests sent by the user.
Response Format: Returns a ConnectionsDTO with lists of user IDs for each category.

3. Unit Tests
Controller Tests (ConnectionControllerTest):
Tests for getConnections, sendConnectionRequest, acceptConnectionRequest, rejectConnectionRequest, and disconnect.
Covers success cases retrieving connections, sending requests and errors like self-requests, unauthorized, invalid states

Service Tests (ConnectionServiceTest):
Tests getConnections for active, pending incoming, and outgoing connections.
Tests sendConnectionRequest for new connections, self-requests, and duplicates.
Tests acceptConnectionRequest, rejectConnectionRequest, and disconnect for success and edge cases not pending, unauthorized

Repository Tests (ConnectionRepositoryTest):
Tests queries like findConnectionsByUserId, existsConnectionBetween, and hasPendingConnectionRequest. database interactions for connection retrieval and state checks.

### CODE REVIEW

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        Long currentUserId = securityUtils.getCurrentUserId(authentication);
+        ConnectionsDTO connections = connectionService.getConnections(currentUserId);
+        return ResponseEntity.ok(connections);
+    }
+
+    @PostMapping("/requests/{userId}")
+    public ResponseEntity<?> sendConnectionRequest(@PathVariable Long userId, Authentication authentication) {
+        Long currentUserId = securityUtils.getCurrentUserId(authentication);
+        connectionService.sendConnectionRequest(currentUserId, userId);
+        return ResponseEntity.ok().build();
```
Comment (karlromets) [inline back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:42]:
```suggestion
        return ResponseEntity.status(HttpStatus.CREATED).build();
```

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    @GetMapping("")
+    public ResponseEntity<ConnectionsDTO> getConnections(Authentication authentication) {
+        Long currentUserId = securityUtils.getCurrentUserId(authentication);
+        ConnectionsDTO connections = connectionService.getConnections(currentUserId);
+        return ResponseEntity.ok(connections);
+    }
+
+    @PostMapping("/requests/{userId}")
+    public ResponseEntity<?> sendConnectionRequest(@PathVariable Long userId, Authentication authentication) {
```
Comment (karlromets) [inline back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:39]:
```suggestion
    public ResponseEntity<Void> sendConnectionRequest(@PathVariable Long userId, Authentication authentication) {
```

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        Long currentUserId = securityUtils.getCurrentUserId(authentication);
+        connectionService.sendConnectionRequest(currentUserId, userId);
+        return ResponseEntity.ok().build();
+    }
+
+    @PatchMapping("/requests/{requestId}/accept")
+    public ResponseEntity<?> acceptConnectionRequest(@PathVariable Long requestId, Authentication authentication) {
```
Comment (karlromets) [inline back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:46]:
```suggestion
    public ResponseEntity<Void> acceptConnectionRequest(@PathVariable Long requestId, Authentication authentication) {
```

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        Long currentUserId = securityUtils.getCurrentUserId(authentication);
+        connectionService.acceptConnectionRequest(requestId, currentUserId);
+        return ResponseEntity.ok().build();
+    }
+
+    @PatchMapping("/requests/{requestId}/reject")
+    public ResponseEntity<?> rejectConnectionRequest(@PathVariable Long requestId, Authentication authentication) {
```
Comment (karlromets) [inline back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:53]:
```suggestion
    public ResponseEntity<Void> rejectConnectionRequest(@PathVariable Long requestId, Authentication authentication) {
```

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        Long currentUserId = securityUtils.getCurrentUserId(authentication);
+        connectionService.rejectConnectionRequest(requestId, currentUserId);
+        return ResponseEntity.ok().build();
+    }
+
+    @DeleteMapping("/{connectionId}")
+    public ResponseEntity<?> disconnect(@PathVariable Long connectionId, Authentication authentication) {
```
Comment (karlromets) [inline back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:60]:
```suggestion
    public ResponseEntity<Void> disconnect(@PathVariable Long connectionId, Authentication authentication) {
```

**Review by karlromets — CHANGES_REQUESTED** [review]
Testid vajavad ka siis update + new praegu failivad

```
[INFO] Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.583 s -- in com.matchme.srv.types.TypesTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   UserQueryServiceTests$GetCurrentUserDTOTests.getCurrentUserDTO_InvalidUser_ThrowsException:213 Multiple Failures (1 failure)
        org.mockito.exceptions.misusing.InvalidUseOfMatchersException: 
Invalid use of argument matchers!
2 matchers expected, 1 recorded:
[ERROR]   ConnectionControllerIntegrationTest.getRecommendations_ShouldReturnMatchingRecommendations_WhenDataExists » IllegalState App
[ERROR]   ConnectionControllerIntegrationTest.getRecommendations_ShouldReturnNotFound_WhenNoMatchesExist » IllegalState Failed to load ApplicationCon
[ERROR]   UserQueryServiceTests$GetCurrentUserDTOTests.getCurrentUserDTO_ValidUser_CallsAccessValidation:236 » PotentialStubbingProblem 
Strict stubbing argument mismatch. Please check:
 - this invocation of 'toCurrentUserResponseDTO' method:
    userDtoMapper.toCurrentUserResponseDTO(
    User(id=1, email=null),
    false
);
    -> at com.matchme.srv.service.user.UserQueryService.getCurrentUserDTO(UserQueryService.java:47)
 - has following stubbing(s) with different arguments:
    1. userDtoMapper.toCurrentUserResponseDTO(
    User(id=1, email=null),
    true
);
      -> at com.matchme.srv.service.user.UserQueryServiceTests$GetCurrentUserDTOTests.getCurrentUserDTO_ValidUser_CallsAccessValidation(UserQueryServiceTests.java:233)
Typically, stubbing argument mismatch indicates user mistake when writing tests.
Mockito fails early so that you can debug potential problem easily.
However, there are legit scenarios when this exception generates false negative signal:
  - stubbing the same method multiple times using 'given().will()' or 'when().then()' API
    Please use 'will().given()' or 'doReturn().when()' API for stubbing.
  - stubbed method is intentionally invoked with different arguments by code under test
    Please use default or 'silent' JUnit Rule (equivalent of Strictness.LENIENT).
For more information see javadoc for PotentialStubbingProblem class.
[ERROR]   UserQueryServiceTests$GetCurrentUserDTOTests.getCurrentUserDTO_ValidUser_ReturnsDTO:191 » PotentialStubbingProblem 
Strict stubbing argument mismatch. Please check:
 - this invocation of 'toCurrentUserResponseDTO' method:
    userDtoMapper.toCurrentUserResponseDTO(
    User(id=1, email=null),
    false
);
    -> at com.matchme.srv.service.user.UserQueryService.getCurrentUserDTO(UserQueryService.java:47)
 - has following stubbing(s) with different arguments:
    1. userDtoMapper.toCurrentUserResponseDTO(
    User(id=1, email=null),
    true
);
      -> at com.matchme.srv.service.user.UserQueryServiceTests$GetCurrentUserDTOTests.getCurrentUserDTO_ValidUser_ReturnsDTO(UserQueryServiceTests.java:187)
Typically, stubbing argument mismatch indicates user mistake when writing tests.
Mockito fails early so that you can debug potential problem easily.
However, there are legit scenarios when this exception generates false negative signal:
  - stubbing the same method multiple times using 'given().will()' or 'when().then()' API
    Please use 'will().given()' or 'doReturn().when()' API for stubbing.
  - stubbed method is intentionally invoked with different arguments by code under test
    Please use default or 'silent' JUnit Rule (equivalent of Strictness.LENIENT).
For more information see javadoc for PotentialStubbingProblem class.
[ERROR]   UserQueryServiceTests$GetCurrentUserDTOTests.shouldReturnEmailSinceRequestedCurrentUserId:138 » PotentialStubbingProblem 
Strict stubbing argument mismatch. Please check:
 - this invocation of 'toCurrentUserResponseDTO' method:
    userDtoMapper.toCurrentUserResponseDTO(
    User(id=1, email=null),
    false
);
    -> at com.matchme.srv.service.user.UserQueryService.getCurrentUserDTO(UserQueryService.java:47)
 - has following stubbing(s) with different arguments:
    1. userDtoMapper.toCurrentUserResponseDTO(
    User(id=1, email=null),
    true
);
      -> at com.matchme.srv.service.user.UserQueryServiceTests$GetCurrentUserDTOTests.shouldReturnEmailSinceRequestedCurrentUserId(UserQueryServiceTests.java:135)
Typically, stubbing argument mismatch indicates user mistake when writing tests.
Mockito fails early so that you can debug potential problem easily.
However, there are legit scenarios when this exception generates false negative signal:
  - stubbing the same method multiple times using 'given().will()' or 'when().then()' API
    Please use 'will().given()' or 'doReturn().when()' API for stubbing.
  - stubbed method is intentionally invoked with different arguments by code under test
    Please use default or 'silent' JUnit Rule (equivalent of Strictness.LENIENT).
For more information see javadoc for PotentialStubbingProblem class.
[INFO] 
[ERROR] Tests run: 186, Failures: 1, Errors: 5, Skipped: 0
```

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java (hunk 2 above)
Comment (K444RU) [reply back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:39]: Changes made: 


```java
    @PostMapping("/requests/{userId}")
    public ResponseEntity<Void> sendConnectionRequest(@PathVariable Long userId, Authentication authentication) {
        Long currentUserId = securityUtils.getCurrentUserId(authentication);
        connectionService.sendConnectionRequest(currentUserId, userId);
        return ResponseEntity.ok().build();
    }```

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java (hunk 1 above)
Comment (K444RU) [reply back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:42]: Changes made:

```java
    @PostMapping("/requests/{userId}")
    public ResponseEntity<Void> sendConnectionRequest(@PathVariable Long userId, Authentication authentication) {
        Long currentUserId = securityUtils.getCurrentUserId(authentication);
        connectionService.sendConnectionRequest(currentUserId, userId);
        return ResponseEntity.status(HttpStatus.CREATED).build();    
    }

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java (hunk 3 above)
Comment (K444RU) [reply back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:46]: Requested changes made

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java (hunk 4 above)
Comment (K444RU) [reply back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:53]: Requested changes made

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java (hunk 5 above)
Comment (K444RU) [reply back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:60]: Requested changes made

File: back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        Long currentUserId = securityUtils.getCurrentUserId(authentication);
+        connectionService.disconnect(connectionId, currentUserId);
+        return ResponseEntity.ok().build();
```
[the remark this answers is not in the document]
Comment (K444RU) [reply back-end/src/main/java/com/matchme/srv/controller/ConnectionController.java:63]: Requested changes made
